fcrepo4/fcrepo4
# Rewrite PATCH request repository URIs to internal IDs

**JIRA Ticket**: https://jira.lyrasis.org/browse/FCREPO-3311

# What does this Pull Request do?

We use the HttpRdfService on PUT/POST requests to ensure RDF has URIs for items in the repository converted to internal IDs. But for Sparql we didn't do this. So while the RDF appears correct the repository doesn't realize it is referencing an item inside the repository.

This adds an "UpdateVisitor" to generate a new Sparql request based on the old one but with any in domain URIs converted to internal IDs

# How should this be tested?

I'm not sure if this is testable except by checking the actual RDF stored in the OCFL objects.

1. Create a RDFSource object.
     `curl -i -ufedoraAdmin:fedoraAdmin -XPOST http://localhost:8080/rest `
1. Patch the resource with some internal IDs.
     `curl -ufedoraAdmin:fedoraAdmin -i -XPATCH -d"INSERT DATA { <http://localhost:8080/rest/resource1> <http://example.org/some/predicate> <http://localhost:8080/rest/someOtherResource> }" -H"Content-type: application/sparql-update" (the URL created above)`
1. Locate and check the `fcr-container.nt` file in the OCFL storage root. It should be in the `v2` directory.

Before the PR it will look like
```
<http://localhost:8080/rest/resource1> <http://example.org/some/predicate> <http://localhost:8080/rest/someOtherResource> .
```
and after it should look like 
```
<info:fedora/resource1> <http://example.org/some/predicate> <info:fedora/someOtherResource> .
```

# Interested parties
@fcrepo4/committers
